text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+4. If the `.out` suffix is not present, the contents of the file must match `<format>`.
+
+Additionally, each hook file may be accompanied by an additional file of the same name, but with the suffix `.buildpacks`.
+This file contains a newline-separated list of buildpack IDs.
+If present, the hook is ignored unless one of the buildpacks in the list is selected during the buildpack detection phase.
+
+Hook files are evaluated (either read or executed) after the detection phase.
+All Dockerfiles are applied to base images after the hooks are evaluated and before the buildpack build phase.
+The resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.
+
+Executable hook files must not modify the app directory when executed and may be evaluated in parallel.
+However, the resulting Dockerfiles must not be applied in parallel to the same base image.
+
+If an executable hook exits with a non-zero status value, the build fails.
+If an executable hook exits with a zero status value and no output, the hook is ignored. 
+Directories are ignored.
+Files at the top-level of `/cnb/hooks.d/` that do not match the specified file name format result in a build failure.
+
+Both Dockerfiles must accept `base_image` and `build_id` args.
+The `base_image` arg allows the lifecycle to specify the original base image.
+The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+
+#### Example: App-specified Dockerfile Hook
+
+This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/hooks.d/app.build.Dockerfile.out`
+```
+#!/bin/sh
+cat build.Dockerfile
+```
+##### `/cnb/hooks.d/app.run.Dockerfile.out`
+```
+#!/bin/sh
+cat run.Dockerfile
+```
+
+#### Example: RPM Dockerfile Hook (app-based)
+
+This example hook would allow a builder to install RPMs for each language runtime, based on the app directory.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+##### `/cnb/hooks.d/app.Dockerfile.out`
+```
+#!/bin/sh
+[[ -f Gemfile.lock ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-ruby
+[[ -f package.json ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-node
```

Comment:
As such checks (to detect if a language/runtime is present within the app's project to be build) has already been done during the `detect` phase, why don't we simply pass a METADATA, env var or yaml/toml file to the hook executable instead of doing again the same check ?

Reply:
> has already been done during the detect phase

As mentioned above, that would couple buildpacks to hooks. Some buildpacks would become entirely dependent on hooks. For a given app and set of buildpacks, it should be possible to create a working builder that doesn't require hooks (since hooks will be slower and break rebase in many cases)

Reply:
> As mentioned above, that would couple buildpacks to hooks. Some buildpacks would become entirely dependent on hooks.

No. As the hook could rely on the METADATA - https://github.com/buildpacks/spec/blob/main/platform.md#detector coming from the `detect` phase, I dont see why you say that we will create a strong dependency with the `build` phase to be executed of a buildpack. such METADATA could simply provide as additional information: JDK11, .... and then the hook will calculate what it should include within the run and/or build Dockerfile generated

Reply:
That metadata (`require`s) has to come from other buildpacks -- that's the coupling I'm worried about.

That said, I'm coming around to this idea, given the observation here: https://github.com/buildpacks/rfcs/pull/173#discussion_r675625620

I'll propose some changes for next week, appreciate the feedback :)